### PR TITLE
Add Condition Variable and 64bit times to library

### PR DESCRIPTION
**Pull request by erlingrj**
Modeled after @sberkun proposal

### CODE REVIEW

**Comment by erlingrj** [comment]
@sberkun would you have the time to look at this?

File: programs/lib/flexpret_cond.c [hunk 1]
```diff
@@ -0,0 +1,50 @@
+#include <flexpret.h>
+#include "flexpret_cond.h"
+
+fp_ret_t cond_wait(cond_t * cond) {
+    int hartid = read_hartid();
+    cond->waiting[hartid] = true;
+    lock_release(cond->lock);
+    while(cond->waiting[hartid]) {}
```
Comment (sberkun) [inline programs/lib/flexpret_cond.c:8]: not sure if `bool waiting[NUM_THREADS];` needs to be marked as `volatile` to keep the compiler from optimizing this loop out

File: programs/tests/c-tests/time/time.c [hunk 2]
```diff
@@ -0,0 +1,21 @@
+#include <stdint.h>
```
Comment (sberkun) [inline programs/tests/c-tests/time/time.c:1]: would be nice to test the edge cases in `rdtime64`, but not sure how to accomplish since that would require simulating on the order of 2^32 cycles

**Review by sberkun — APPROVED** [review]
overall looks good to me!

File: programs/lib/include/flexpret_time.h [hunk 3]
```diff
@@ -32,5 +29,36 @@ static inline void delay_for(unsigned int duration_ns)
...
   delay_until(now_ns + duration_ns);
 }
+/**
+ * @brief Read out a 64 bit timestamp. Since it is done with two read operations
+ * we must handle a potential wrapping event where we read the lower bits BEFORE the wrap
+ * and the higher bits AFTER the wrap.
+ * 
+ * @return uint64_t 
+ */
+static inline uint64_t rdtime64()
```
Comment (sberkun) [inline programs/lib/include/flexpret_time.h:39]: I think the implementation is easier if you read high-low-high instead of low-high-low:
```c
static inline uint64_t rdtime64() 
{
  uint32_t hi_pre = read_csr(CSR_TIMEH);
  uint32_t lo = rdtime();
  uint32_t hi_post = read_csr(CSR_TIMEH);

  // in the event of a wraparound, lo was read either before the wrap or after 
  // the wrap. If lo > 2^31, then lo was read before the wrap. Otherwise, lo
  // was read after the wrap. 
  uint32_t hi = (lo >> 31) ? hi_pre : hi_post;
  return (uint64_t) hi << 32 | lo;
}
```

Comment (erlingrj) [reply programs/lib/include/flexpret_time.h:39]: I dont understand this proposal completely. `  uint32_t hi = (lo >> 31) ? hi_pre : hi_post;` says that if the MSb of `lo` is 1 then we should use `hi_pre`, and if it is 0 we should use `hi_post`. This is not safe considering SRTT and interrupts. What if the SRTT is delayed for a long period after reading, hi_pre and lo, during that period we could experience a wrap. Then this logic would pick `lo` and `hi_post` which would be wrong

File: programs/lib/flexpret_cond.c (hunk 1 above)
Comment (erlingrj) [reply programs/lib/flexpret_cond.c:8]: Good point. I am actually also not sure. But lets do it for good measure!

File: programs/lib/include/flexpret_time.h (hunk 3 above)
Comment (erlingrj) [reply programs/lib/include/flexpret_time.h:39]: Actually, my implementation also has some assumptions, namely that there only ever will be a single wrap between the reading of the lower and higher bits. But with SRTT you don't really know. It is also a reasonable question to ask whether a timestamp has any value with that type of uncertainty though.

File: programs/tests/c-tests/time/time.c (hunk 2 above)
Comment (erlingrj) [reply programs/tests/c-tests/time/time.c:1]: We could allow writes to the counter register. Might be useful for clock synchronization. But might be very dangerous also

**Review by erlingrj — COMMENTED** [review]
Thanks for feedback @sberkun, I will address this soon

File: programs/lib/include/flexpret_time.h (hunk 3 above)
Comment (sberkun) [reply programs/lib/include/flexpret_time.h:39]: We can define correctness of `rdtime64()` as following:
 - if the function runs for under 2^32 nanoseconds, it returns the current time at the beginning of the function (when `lo_pre` is read)
 - if the function runs for over 2^32 nanoseconds, it returns some time that occurred when the function was running

The second condition ensures that subsequent calls to `rdtime64()` are guaranteed to be monotonically increasing. I'm pretty sure your implementation satisfies these two requirements, regardless of how many wraps occur `lo_pre` and `hi`.

The implementation I provided satisfies the 1st requirement, but not the 2nd.
